### DISCUSSION TITLE
Generate Client-Traits for services?

### DISCUSSION THREAD

**Discussion by buehler** [tokio-rs/prost#773] [Q&A]
Hey there

I'm using prost/tonic for several projects now. However, I'm currently maintaining an open-source library and I'm wondering if it is possible to generate client implementation traits from proto files.

The use-case would be to create a builder pattern for a grpc client but then return an `impl MyClient` instead of a concrete struct type.

This, to enable attaching interceptors and setting other options.

It would be nice to have a trait for all methods in a service.

Regards

**Comment by LucioFranco** [comment]
There is the `Service` trait that you can use to generate clients/servers etc.

**Reply by buehler** [reply]
What do you mean by `Service` trait?

Is there a trait that the client/server does conform to?

Basically, like the implementation on the server side which is done via async trait. I'd like to have a trait for the client side.

Idea being that I can make a builder struct that sets some options and then returns an `impl MyClient`.

E.g.:
```protobuf
service HelloService {
  rpc World(Emtpy) returns (Empty);
}
```

This would generate the client code and a trait like:
```rust
pub trait HelloServiceClientTrait {
  async fn world(e: Empty) -> io::FutureResultWhateverTheTypeWas<Empty>;
}

impl HelloServiceClientTrait for HelloServiceClient {
  // actual client ipmlementation
}
```

The result of this trait could then be:
```rust
struct MyClientBuilder;

impl MyClientBuilder {
  pub fn build(&self) -> impl HelloServiceClientTrait {
    // ... stuff
  }
}
```

The use-case is here:
- "plain": https://github.com/smartive/zitadel-rust/blob/main/src/api/clients.rs#L52
- inside mod "with access token": https://github.com/smartive/zitadel-rust/blob/main/src/api/clients.rs#L194
- inside mod "with service account": https://github.com/smartive/zitadel-rust/blob/main/src/api/clients.rs#L325

I need three modules with three times the same method to return a special client based on the interceptors that should be attached. It would be nice to have a builder pattern (or maybe someone has a good idea on the topic), that just returns the type that implements the client.

**Reply by buehler** [reply]
Add:
In the referenced lines of code, I'd like to be able to return `ManagementServiceClient` instead of `ManagementServiceClient<InterceptedService<Channel, AccessTokenInterceptor>>`. Then it would be possible to build/create the service client according to the needs of the builder input.

**Reply by LucioFranco** [reply]
So this seems more like a tonic question rather than a prost one. What I was referring too was this https://docs.rs/prost-build/latest/prost_build/trait.ServiceGenerator.html in prost will allow you to generate what ever you want from the proto info.

As for the trait, tonic doesn't support generating traits for clients but instead promotes using different types of channels. Aka if you wanted to mock stuff you would use a mocked in memory channel and swap it for the real channel for prod use. 

You can always write something custom via this prost trait, you can reference the tonic code for how its used.
